### PR TITLE
Goal lines refactoring

### PR DESCRIPTION
**Pull request by jamiecarter7**
**Story card:** None - bug fix

## Because

Needed to refactor monthIndexFromDateKey() function. 
This will be needed in a follow up PR that fixes a bug on the dashed lines of the overdue chart

## This addresses

Refactor complete plus some additional work to simplify the goal lines functionality and make it easier to follow

Changes:
- goalPeriodValue() renamed to getThreeMonthAverageAndImprovementRatio()
- improvement ratio is calculated inside the getThreeMonthAverageAndImprovementRatio() rather than passing data around
- simplified check for isEndMonthOfYear() (previously named isLastDateKeysArrayMonthDec)
- improved variable naming
- monthIndexFromDateKey() uses built in javascript function
- calculateGoalUpwards() and calculateGoalDownwards() have parameter names that match the data being passed (monthValue -> threeMonthAverage)
- correction to how the remaining months in the year is calculated: 
```
const monthsRemainingForYear = monthThreeIndex === 11 ? 12 : 11 - monthThreeIndex;
```

## Test instructions

Switch between this branch and the current master branch to check that the figures are the same.
Note: there was a previous miscalculation on the goal lines.
When checking the figures are the same there might be a small chance 'this branch/PR' shows a figure that is 1% less than that shows on the current master. This is correct - the master was +0.0083 higher than it should have been after running the calculations. Figures are rounded so there should be no change unless it crosses to the next integer.

To view the goal lines add the commented line to your reports.js file

``` 
const enabledRegions = {
      IN: ["state"], // add this line <-
      BD: ["organization", "state", "district"],
      ET: ["organization", "state"],
      LK: ["organization"],
    };
```
the line for copying
``` 
IN: ["state"], 
```

### CODE REVIEW

File: app/assets/javascripts/common/reports.js [hunk 1]
```diff
@@ -662,83 +662,69 @@ Reports = function ({
...
   }
 
-  function relativeImprovementRatio(goalMonthIndex) {
-    const defaultRelativeImprovementPercentage = 10;
-    const improvementPercentagePerMonth =
-      defaultRelativeImprovementPercentage / 100 / 12;
-    const monthsRemainingInYear = 12 - (goalMonthIndex % 11); // dec is full year (0 index)
-    return improvementPercentagePerMonth * monthsRemainingInYear;
+  function calculateGoalDownwards(threeMonthAverage, improvementRatio) {
+    const goal = threeMonthAverage - threeMonthAverage * improvementRatio;
```
Comment (PriyangaPKini) [inline app/assets/javascripts/common/reports.js:726]: [nitpick] It's a good practice to add brackets around the expression that is to be evaluated first
```suggestion
    const goal = threeMonthAverage - (threeMonthAverage * improvementRatio);
```

File: app/assets/javascripts/common/reports.js [hunk 2]
```diff
@@ -662,83 +662,69 @@ Reports = function ({
...
+    const monthlyRelativeImprovement = annualRelativeImprovement / 12;
+    const monthsRemainingForYear = monthThreeIndex === 11 ? 12 : 11 - monthThreeIndex; // dec = 12, jan = 11... nov = 1
+    return monthlyRelativeImprovement * monthsRemainingForYear;
   }
 
-  function calculateGoalDownwards(monthValue, improvementRatio) {
-    const goal = monthValue - monthValue * improvementRatio;
-    return Math.floor(goal);
+  function calculateGoalUpwards(threeMonthAverage, improvementRatio) {
+    const goal = threeMonthAverage + (100 - threeMonthAverage) * improvementRatio;
```
Comment (PriyangaPKini) [inline app/assets/javascripts/common/reports.js:721]:
```suggestion
    const goal = threeMonthAverage + ((100 - threeMonthAverage) * improvementRatio);
```

**Review by PriyangaPKini — COMMENTED** [review]
Left some comments. Please feel free to merge the PR after resolving the comments

File: app/assets/javascripts/common/reports.js [hunk 3]
```diff
@@ -662,83 +662,69 @@ Reports = function ({
...
       return calculateGoalDownwards(threeMonthAverage, improvementRatio);
     }
     return calculateGoalUpwards(threeMonthAverage, improvementRatio);
   }
 
-  function goalPeriodValue(periodValues) {
-    const dateKeysArray = Object.keys(periodValues);
+  function getThreeMonthAverageAndImprovementRatio(periodValues) {
+    const dateKeys = Object.keys(periodValues);
+    const decemberKeys = dateKeys.filter((item) => item.includes("Dec"));
```
[the remark this answers is not in the document]
Comment (jamiecarter7) [reply app/assets/javascripts/common/reports.js:675]: changed to evaluationMonth from all instances of Dec
Dec is also now its own `const evaluationMonth = "Dec"`

**Pull request merged**